File: sdk-typescript/test/iwfcompat/timer_test.ts
// Copyright (c) 2026 Super Durable, Inc.
//
// Licensed under the Super Durable Source License 1.0.
// You may not use this file except in compliance with the License.
// See the LICENSE file in the repository root.
//
// SPDX-License-Identifier: LicenseRef-Super-Durable-1.0

import { StepExecutionId, type Client } from "../../src/index.js";

import * as flows from "./iwf_flows.js";

export async function compileTimerAndStepWait(client: Client): Promise<void> {
  await client.startFlow(flows.TIMER, "timer", 1);
  await client.waitForStepCompletion("timer", StepExecutionId.of("TimerStep"), 10_000);
  await client.waitForFlow("timer");
}

File: sdk-rust/crates/dex-sdk/tests/integ/timer_test.rs
// Portions of this file are derived from indeedeng/iwf-java-sdk.
// Those portions are licensed under the Apache License, Version 2.0.
// See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
//
// Modifications Copyright (c) 2026 Super Durable, Inc.
//
// Modifications are licensed under the Super Durable Source License 1.0.
// Third-Party Materials remain under the Apache License, Version 2.0.
// See LICENSE and LEGACY_NOTICES.md.

use std::time::{Duration, Instant};

use dex_sdk::{Client, Registry, SdkResult, StepExecutionId};

use crate::support::{DexDevTestEnvironment, flow_id};
use crate::timer_workflow::TimerWorkflow;

#[test]
#[ignore = "requires dexcli dev"]
fn test_basic_timer_workflow() {
    let environment = DexDevTestEnvironment::start(Registry::new().register(TimerWorkflow::new()));
    let workflow = TimerWorkflow::new();
    let flow_id = flow_id("basic-timer");
    let started_at = Instant::now();
    environment
        .client
        .start_flow(&workflow, &flow_id, 5)
        .expect("start timer Flow");
    environment
        .client
        .wait_for_step_completion(
            &flow_id,
            StepExecutionId::of(&workflow.start),
            Duration::from_secs(10),
        )
        .expect("wait for Timer Step");
    environment
        .client
        .wait_for_flow_with_timeout(&flow_id, Duration::from_secs(30))
        .expect("complete timer Flow");
    let elapsed = started_at.elapsed();
    assert!(
        (Duration::from_secs(4)..=Duration::from_secs(7)).contains(&elapsed),
        "actual duration: {elapsed:?}"
    );
}

#[allow(dead_code)]
fn compile_timer_and_step_wait(client: &Client) -> SdkResult<()> {
    let workflow = TimerWorkflow::new();
    client.start_flow(&workflow, "timer", 1)?;
    client.wait_for_step_completion(
        "timer",
        StepExecutionId::of(&workflow.start),
        Duration::from_secs(10),
    )?;
    let _ = client.wait_for_flow("timer")?;
    Ok(())
}
